modules/gapi/src/streaming/onevpl/demux/async_mfp_demux_data_provider.cpp

```diff
@@ -0,0 +1,829 @@
+        handler->GetCurrentMediaType(&media_type_tmp);
+        if (!media_type_tmp) {
+            GAPI_LOG_WARNING(nullptr, "[" << this << "] " <<
+                                      "Cannot determine media type for stream by index: " <<
+                                      stream_index);
+            continue;
+        }
+
+        ComPtrGuard<IMFMediaType> media_type = createCOMPtrGuard(media_type_tmp);
+        GUID subtype;
+        if (SUCCEEDED(media_type->GetGUID(MF_MT_SUBTYPE, &subtype))) {
+            GAPI_LOG_DEBUG(nullptr, "[" << this << "] " <<
+                                    "video type: " << GetGUIDNameConst(subtype));
+
+            std::string is_codec_supported("unsupported, skip...");
+            try {
+                codec = convert_to_mfx_codec_id(subtype);
+                is_stream_selected = true;
+                is_codec_supported = "selected!";
+            } catch (...) {}
+
+            GAPI_LOG_INFO(nullptr, "[" << this << "] " <<
+                          " by URI: " << file_path <<
+                          ", video stream index: " << stream_index <<
+                          ", codec: " << GetGUIDNameConst(subtype) <<
+                          " - " << is_codec_supported)
+        } else {
+            GAPI_LOG_WARNING(nullptr, "[" << this << "] " <<
+                                      "Cannot media GUID subtype for stream by index: " <<
+                                      stream_index);
+            continue;
+        }
+    }
+
+    if (!is_stream_selected) {
+        const auto &supported_codecs = get_supported_mfx_codec_id();
+        std::string ss;
+        for (mfxU32 id : supported_codecs) {
+            ss += mfx_codec_id_to_cstr(id);
+            ss += ", ";
+        }
+        if (!ss.empty()) {
+            ss.erase(ss.size() - 2, 2);
+        }
+
+        GAPI_LOG_WARNING(nullptr, "[" << this << "] "
+                         "couldn't find video stream with supported params, "
+                         "expected codecs: " << ss);
+        throw DataProviderUnsupportedException("couldn't find supported video stream");
+    }
+
+    GAPI_LOG_DEBUG(nullptr, "[" << this << "] " <<
+                            "is creating media source");
+    IMFSourceReader *source_reader_tmp = nullptr;
+    hr = MFCreateSourceReaderFromMediaSource(source.get(), attributes.get(),
+                                             &source_reader_tmp);
+    if (FAILED(hr)) {
+        throw DataProviderSystemErrorException(HRESULT_CODE(hr), "Cannot create MFCreateSourceReaderFromMediaSource");
+    }
+    source_reader = createCOMPtrGuard(source_reader_tmp);
+
+    GAPI_LOG_DEBUG(nullptr, "[" << this << "] " <<
+                            "created IMFSourceReader: " << source_reader);
+
+    if (SUCCEEDED(hr)) {
+        // Ask for the first sample.
+        request_next(hr, 0, 0);
+    } else {
+        throw DataProviderSystemErrorException(HRESULT_CODE(hr), "Cannot ReadSample");
```

Comment:
"Cannot ReadSample"" or something which is already handled on the 457 line?
```
if (FAILED(hr)) {
        throw DataProviderSystemErrorException(HRESULT_CODE(hr), "Cannot create MFCreateSourceReaderFromMediaSource");
```

This is confusing. How can we reach this `else` statement?